Ensure there is one running static pod with the same full name

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR ensures there is no more than one running static pod with the same full name so that static pods are able to be updated/deleted gracefully.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #97722

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Respect grace period when updating static pods.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

gjkim42 committed:
NodeConformance: Respect grace period when updating static pod

gjkim42 committed:
Ensure there is one running static pod with the same full name

gjkim42 commented:
/retest

gjkim42 commented:
/assign @smarterclayton 

Made an alternative to #104031

gjkim42 commented:
/retest

I think kubelet-serial tests are broken

https://prow.k8s.io/?repo=kubernetes%2Fkubernetes&job=pull-kubernetes-node-kubelet-serial

gjkim42 commented:
/sig node
/priority important-soon

rphillips commented:
@gjkim42 this PR is not fixing the issue for me... I am reproducing it with @liggitt's steps outlined: https://github.com/kubernetes/kubernetes/issues/104648#issue-981968741

gjkim42 commented:
@rphillips 

This PR is not directly relevant to #104648.
This PR fixes #97722 - kubelet allows two running static pods having the same full name and different uids which is not desirable.

#104648 is that kubelet ignores a recreation operation because an old static pod and a new static pod (with the same contents) have the same uid.
If we want to address #104648 by generating a new uid for the new static pod(which has the same full name as an old one), we also have to fix #97722. Without this PR, making two static pods having the same name and different uids is not safe.

endocrimes commented on pkg/kubelet/pod_workers.go:
```diff
@@ -664,6 +684,18 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
 
+// isConflictedPod returns true when there is a known pod with the same
+// fullname as pod
+func (p *podWorkers) isConflictedPod(pod *v1.Pod) bool {
```
```suggestion
func (p *podWorkers) isConflictingStaticPod(pod *v1.Pod) bool {
```

Lets help future folks and reduce the scope of the name a little

endocrimes commented on pkg/kubelet/pod_workers.go:
```diff
@@ -765,12 +797,18 @@ func (p *podWorkers) managePodLoop(podUpdates <-chan podWork) {
 			klog.ErrorS(err, "Error syncing pod, skipping", "pod", klog.KObj(pod), "podUID", pod.UID)
 
 		case update.WorkType == TerminatedPodWork:
+			klog.V(4).InfoS("Processing pod event done", "pod", klog.KObj(pod), "podUID", pod.UID, "updateType", update.WorkType)
```
The processing hasn't completed yet, so this log becomes potentially quite misleading.

endocrimes commented on pkg/kubelet/pod_workers.go:
```diff
@@ -819,6 +857,44 @@ func (p *podWorkers) acknowledgeTerminating(pod *v1.Pod) PodStatusFunc {
 	return nil
 }
 
+// queueReservedWork queues a reserved work having the same full name as pod
```
Can you add the why here? - It won't necessarily be immediately obvious to future readers of this code.

endocrimes commented on pkg/kubelet/pod_workers.go:
```diff
@@ -819,6 +857,44 @@ func (p *podWorkers) acknowledgeTerminating(pod *v1.Pod) PodStatusFunc {
+// queueReservedWork queues a reserved work having the same full name as pod
+// and completes the previous pod. It returns true if there is a reserved work
+// else returns false
+func (p *podWorkers) queueReservedWork(pod *v1.Pod) bool {
```
```suggestion
func (p *podWorkers) queueWorkForConflictedStaticPod(pod *v1.Pod) bool {
```

This currently only applies to a small set of pods, lets avoid having a generic name that's easy to piggyback onto in the future without clear intent

endocrimes commented on pkg/kubelet/pod_workers.go:
```diff
@@ -620,12 +626,26 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
+					p.podSyncStatuses[pendingPod.UID].finished = true
+				}
+			}
+			klog.InfoS("Reserve a podWork for the conflicted pod", "pod", klog.KObj(pod), "podUID", pod.UID)
```
```suggestion
			klog.InfoS("Reserving work to update the conflicted pod", "pod", klog.KObj(pod), "podUID", pod.UID)
```

endocrimes commented on pkg/kubelet/pod_workers.go:
```diff
@@ -765,12 +797,18 @@ func (p *podWorkers) managePodLoop(podUpdates <-chan podWork) {
+				if start := update.Options.StartTime; !start.IsZero() {
+					metrics.PodWorkerDuration.WithLabelValues("terminated").Observe(metrics.SinceInSeconds(start))
+				}
+				continue
```
Add a comment explaining why we maintain the loop here?

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -620,12 +626,26 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
+				pendingPod := pendingWork.Options.Pod
+				if pendingPod.UID != uid {
+					klog.V(4).InfoS("Mark an outdated pod as finished", "pod", klog.KObj(pod), "podUID", pod.UID)
+					p.podSyncStatuses[pendingPod.UID].finished = true
```
This isn't safe because there is no guarantee pendingPod.UID doesn't collide with a running pod (and any running pod can't be terminated early).

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -819,6 +857,44 @@ func (p *podWorkers) acknowledgeTerminating(pod *v1.Pod) PodStatusFunc {
+	p.podSyncStatuses[work.Options.Pod.UID].working = true
+	podUpdates <- work
+
+	delete(p.podUpdates, pod.UID)
```
Since this is identical to complete terminating, I really didn't expect this to be here (but instead expected the chunk in managePodLoop to invoke this method, then invoke completeTerminating)

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -765,12 +797,18 @@ func (p *podWorkers) managePodLoop(podUpdates <-chan podWork) {
 
 		case update.WorkType == TerminatedPodWork:
+			klog.V(4).InfoS("Processing pod event done", "pod", klog.KObj(pod), "podUID", pod.UID, "updateType", update.WorkType)
+			if p.queueReservedWork(pod) {
```
The usage of a method that returns a bool that is not named 'has*|should*' is a bit unusual.  I'm confused though - we need to wake up *another* worker (because the queued work isn't guaranteed to have our uid, in fact is very likely not to), so I expected this chunk to simply queue an action on another pod's work queue.  That implies that we start one work queue for every incoming uid, and for a given uid we may have updates for that uid be "no-ops" until the previous duplicate is freed.

Also, many pods can attempt to use the same full name, so I expected the reserved work for a pod name to be a list, not a single entry.  

Finally, if a pod is blocked long enough on shutdown I think a reason is going to have to be provided to pending pod status.  I'm a little uncomfortable about having that in this method, so it might simply be a new sync method (`syncWaitingPod`) or something provided to this method.

gjkim42 replied:
> Finally, if a pod is blocked long enough on shutdown I think a reason is going to have to be provided to pending pod status. I'm a little uncomfortable about having that in this method, so it might simply be a new sync method (`syncWaitingPod`) or something provided to this method.

@smarterclayton 
Can you please elaborate on this? 
I am not sure where to call the method that sets a reason for the pending pod status.
Do we want to give such work to the pending pod's goroutine?

gjkim42 commented:
@smarterclayton @rphillips 

It seems that changing a static pod file is similar to force deleting a pod and making a new one immediately.
(They will make `podWorkers` have two pods with the same full name and the different uids.)

So, I think this change may change the behavior of the force deleting static pods.

i.e.
Without this change, force deleting a static pod will have kubelet recreate the static pod immediately.
However, with this change, force deleting a static pod will have kubelet wait for the old static pod to be deleted gracefully.

Not sure this is ok. Need your advice.

gjkim42 commented:
/retest

smarterclayton commented:
> However, with this change, force deleting a static pod will have kubelet wait for the old static pod to be deleted gracefully.

The reason to allow static UID on static pods was mostly to ensure inline changes of pods didn't involve tearing them down, and possibly also to guarantee that you don't have two instances of the same pod running at the same time when using tsatic pods (I.e. if you update the static pod you want the old one stopped before the new one starts).

I believe we are working towards the following:

1. A user can ensure a static pod is updated but that the old instance is stopped before the new one is started (by having the same pod name / namespace)
2. A user can update a static pod with an incremental change efficiently.

Currently, this change would make 1 correct.  #104712 by itself would break 1.  The two changes together  would  makes 2 worse (because after that change ALL static pods involve full shutdown and full restart), but would potentially prevent users from making unsupported changes to pods accidentally (i.e. adding a new container to a pod).  I'm not sure yet whether we want to do #104712, but we definitely need to make this change.

gjkim42 commented:
> 1. A user can ensure a static pod is updated but that the old instance is stopped before the new one is started (by having the same pod name / namespace)

What I was trying to say is that making 1 correct would make a new side effect - we will not support force deleting static pods.

gjkim42 commented:
@smarterclayton 

> 2\. A user can update a static pod with an incremental change efficiently (add env var, change container image) without recreating the entire pod

@yujuhong (the original author of this implementation) said that
https://github.com/kubernetes/kubernetes/issues/6580#issuecomment-91079862

Updating static pods without recreating them has not been supported for a long ago (about many years).

249043822 commented on pkg/kubelet/pod_workers.go:
```diff
@@ -620,12 +629,18 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
 	var exists bool
 	if podUpdates, exists = p.podUpdates[uid]; !exists {
+		fullname := kubecontainer.GetPodFullName(pod)
+		if p.isConflictedStaticPod(pod) {
```
only the same static pod， while it's SYNC work, it still enters this condition, that's not correct. so this condition maybe not enough

gjkim42 replied:
Can you elaborate on this? Which condition do you think we need to add?

We are trying to ensure that there is only one running static pod with the same full name.

249043822 replied:
this condition is if `p.knownPodFullnames[fullName]` already has the pod full name, it considers as conflict,  but with the same static pod, it still enter the process periodly for it's sync interval, so `p.reservedPodUIDsFullnames[fullname]` may store many unuseful uids.

gjkim42 replied:
> but with the same static pod, it still enter the process periodly for it's sync interval,

I think the same(having the same uid) static pod cannot enter here twice. After a static pod enters here, podUpdates[uid] exists for the same pod, so we can't enter again for the same static pod.

Please let me know if I misunderstand.

249043822 replied:
oh sorry, it's my mistake

249043822 commented on pkg/kubelet/pod_workers.go:
```diff
@@ -637,8 +652,9 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
-	if !status.IsWorking() {
+	// dispatch a request to the pod worker if none are running and the pod is
+	// not reserved
+	if !status.IsWorking() && !status.IsReserved() {
```
Does add ` !status.IsReserved()` here properly? Because if a new static pod with same full name comes,  status.reserved becomes true, then the old pod worker couldn't enter this condtion again, it would hinder managePodLoop

gjkim42 replied:
An old static pod and a new static pod have different uids, and status is retrieved by the uid.
So old pod's `status.IsReserved` would false.

249043822 replied:
Is it possible that while kubelet restart, the reserved becomes creating runtime pod, but the original running pod becomes reserved, but it's containers are still running and managePodLoop stop

gjkim42 replied:
I've inspected this behavior. To summary, yes it is possible but it won't have managePodLoop stop.

When restarting the kubelet, a new static pod that should be reserved can enter the UpdatePod first. (in most cases)

However, there is no config for the old static pod but there is only an orphaned runtime pod that should terminate.
And, an orphaned runtime pod is considered a non-static pod so isConflictedStaticPod returns false for the old static pod and then the old static pod is being terminated.

When restarting kubelet, it is possible that two static pods with the same name can be running for a little while. (we may tolerate this...)
However, there is no deadlock.

gjkim42 commented:
/retest

gjkim42 commented:
/retest

smarterclayton commented:
https://docs.google.com/document/d/1NJKYNgoXZKGS5la4MvGrB42-DNQSFf4JQkEcZRnrDzA/edit# covers some of the implications of the larger set of problems around static pods and restarts.  I think this fix is one of the necessary components in order to provide exclusion when someone doesn't use a UID (which removes the need for most people to use a UID)

smarterclayton commented:
> Updating static pods without recreating them has not been supported for a long ago (about many years).

That comment is no longer correct - we indeed must support valid changes to static pods because we support static pods.

gjkim42 commented:
> https://docs.google.com/document/d/1NJKYNgoXZKGS5la4MvGrB42-DNQSFf4JQkEcZRnrDzA/edit#

I think i don't have an access to read this.

gjkim42 commented:
/retest-required

gjkim42 commented:
@smarterclayton 

It seems that force deleting static pods has not been supported at least since v1.21.0. I suspect that it has not been supported a long ago.

If we do
$ kubectl  delete pods STATIC_POD --force 
then it deletes the corresponding mirror pod, not the static pod.

To summary, this PR will not affect force deleting static pods.

smarterclayton commented:
> To summary, this PR will not affect force deleting static pods in API server.

I was not referring to force deleting static pods in the API server.  I was referring to the act of deleting a static pod file, which is a "force delete" in the same fashion as the apiserver (the pod disappears from configuration).

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -882,6 +907,8 @@ func (p *podWorkers) completeTerminatingRuntimePod(pod *v1.Pod) {
 	delete(p.lastUndeliveredWorkUpdate, pod.UID)
 	delete(p.terminatingStaticPodFullnames, kubecontainer.GetPodFullName(pod))
+
+	p.dispatchReservedStaticPodWork(kubecontainer.GetPodFullName(pod))
```
Setting this up at the end is confusing.  I would simply prefer to spawn a regular pod worker for every pod (even if it's conflicting), and then have the pod manage loop check for conflicts (if it is blocked, simply queue a notification) and also have the termination of a pod wake any components it was flushing.  That avoids the need to have dispatchReservedStaticPodWork and having pod work queues get managed in two places (I would prefer that only UpdatePod queue work, and managePodLoop only requeue others).

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -387,18 +394,20 @@ func newPodWorkers(
+		podSyncStatuses:                map[types.UID]*podSyncStatus{},
+		podUpdates:                     map[types.UID]chan podWork{},
+		lastUndeliveredWorkUpdate:      map[types.UID]podWork{},
+		terminatingStaticPodFullnames:  map[string]struct{}{},
```
This map should be unnecessary if managePodLoop takes ownership of preventing duplicate starts - roughly at the top of the loop:

```
var hasStarted bool
for range podUpdates {
  if !hasStarted {
    if !p.allowPodStart(pod) { // calculates whether the pod can be started
      p.workQueue.Enqueue(pod.UID, /* some resync interval */)
      klog // log that we can't start yet consistent with other log statements in this loop
      continue
    }
    hasStarted = true
  }
}
```

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -626,6 +632,16 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
 		// the channel is empty, so buffer of size 1 is enough.
 		podUpdates = make(chan podWork, 1)
 		p.podUpdates[uid] = podUpdates
+		fullname := kubecontainer.GetPodFullName(pod)
```
This could belong inside isAllowedToStart

smarterclayton replied:
This whole chunk only applies for static pods (which means inside isAllowedToStart it's after the isStaticPod check)

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -697,10 +729,21 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+			if !p.isAllowedToStart(pod) {
+				p.workQueue.Enqueue(pod.UID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
+				p.completeWorkQueueNext(pod.UID)
+				klog.V(4).InfoS("Pod update is reserved", "pod", klog.KObj(pod), "podUID", pod.UID)
```
"Pod cannot start yet"

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -882,6 +922,8 @@ func (p *podWorkers) completeTerminatingRuntimePod(pod *v1.Pod) {
 	delete(p.lastUndeliveredWorkUpdate, pod.UID)
 	delete(p.terminatingStaticPodFullnames, kubecontainer.GetPodFullName(pod))
+
+	p.startNextStaticPod(kubecontainer.GetPodFullName(pod))
```
Why is this needed now?  isAllowedToStart should be able to determine this by checking whether it's the first allowed pod in the list of pending.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -387,18 +391,20 @@ func newPodWorkers(
+		podUpdates:                         map[types.UID]chan podWork{},
+		lastUndeliveredWorkUpdate:          map[types.UID]podWork{},
+		terminatingStaticPodFullnames:      map[string]struct{}{},
+		startedPodUIDsByFullname:           map[string]types.UID{},
```
You only need startedPodUIDsByFullname and can remove terminatingStaticPodFullnames.  IsPodForMirrorPodTerminatingByFullName now must calculate by checking startedPodUIDsByFullname, then checking the podSyncStatus for that UID and see whether it is both terminationRequested and ! terminated.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -626,6 +632,16 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
+		startedUID, started := p.startedPodUIDsByFullname[fullname]
+		if !started || startedUID == uid {
+			// we can start a pod
+			p.startedPodUIDsByFullname[fullname] = uid
```
If this becomes only static pods, it should describe that it only applies to static pods.

gjkim42 commented:
/retest-required

gjkim42 commented:
/test-required

gjkim42 commented:
/retest-required

gjkim42 commented:
@smarterclayton 
I think they are all done! Could you PTAL?

gjkim42 commented:
/retest

gjkim42 commented:
/retest

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -696,11 +710,70 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+			if !staticPodStarted {
+				// try to start static pod
+				func() {
+					p.podLock.Lock()
```
This needs to be a function, and IsStaticPod(pod) should be inside that function.

```
if p.allowPodStart(pod) {
```

```
func allowPodStart(pod) bool {
  if !kubetypes.IsStaticPod(pod) { 
     return true
  }
  ...
  p.podLock.Lock()
  ...
}
```

smarterclayton replied:
In general "if static pod" is an assumption that will change, so main code loops should be shielded from knowing about that.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -734,10 +748,71 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+	}
+
+	p.startedStaticPodsByFullname[fullname] = waitingToStartPods[0]
+	p.waitingToStartStaticPodsByFullname[fullname] = waitingToStartPods[1:]
```
I *think* we need to iterate through waitingToStartPods and remove any uid that is not in podSyncStatus, then take the one in the first position.

Also, can the current UID be started AND in be in the first waitingToStartPods position?  If so, we may also need to exclude that.

I'd like to see unit tests for allowPodStart that capture various uses cases and describe which aren't solved (in this PR).

smarterclayton replied:
(full branch coverage on these methods added for a `Test_allowPodStart`)

gjkim42 replied:
> I'd like to see unit tests for allowPodStart that capture various uses cases and describe which aren't solved (in this PR).

done



> I _think_ we need to iterate through waitingToStartPods and remove any uid that is not in podSyncStatus, then take the one in the first position.
> 
> Also, can the current UID be started AND in be in the first waitingToStartPods position? If so, we may also need to exclude that during iteration.

They may be possible. And I think that it is not desirable to leave podWorkers in an inconsistent state(partially completed pod) at any time except inside of the podLock.

So, fixed podWorkers to be able to complete pod clearly.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -734,10 +748,71 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+
+	waitingToStartPods := p.waitingToStartStaticPodsByFullname[fullname]
+	if len(waitingToStartPods) == 0 {
+		return false
```
If we are not waiting, shouldn't this be true?  "A static pod is allowed to start if there are no waiting static pods with that full name"?

gjkim42 commented:
/retest

gjkim42 commented:
/retest

MadhavJivrajani commented:
/triage accepted

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -936,7 +1031,24 @@ func (p *podWorkers) completeTerminated(pod *v1.Pod) {
 	delete(p.podUpdates, pod.UID)
 	delete(p.lastUndeliveredWorkUpdate, pod.UID)
-	delete(p.terminatingStaticPodFullnames, kubecontainer.GetPodFullName(pod))
+	if fullname, exists := p.podFullnames[pod.UID]; exists {
```
Is this chunk and the following the same?

smarterclayton replied:
Why are you doing the logic to keep waitingToStartStaticPods in `complete*()` vs inside of `allowNextStaticPod()`?  Is there a reason to have it here?

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -375,9 +374,12 @@ type podWorkers struct {
-	// static pod transitions to the killed state.
-	terminatingStaticPodFullnames map[string]struct{}
+	// Tracks all full names for known pods
+	podFullnames map[types.UID]string
```
It would be better to put the full name in the `podSyncStatus` struct vs having another map to maintain here (unless podFullnames has a different set than podSyncStatuses, which would be confusing and likely to cause issues).

gjkim42 commented:
@smarterclayton 
tested multiple updates during a graceful termination and fixed some codes.
fixed to finish not started static pod if it is no longer known.

Could you PTAL?

gjkim42 commented:
/retest

gjkim42 commented:
/retest

gjkim42 commented:
/retest

gjkim42 commented:
/retest

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -665,6 +675,13 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
 		p.podUpdates[uid] = podUpdates
 
+		if kubetypes.IsStaticPod(pod) {
+			fullname := kubecontainer.GetPodFullName(pod)
```
This is in `status.fullname`, shouldn't need to recalculate it.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -665,6 +675,13 @@ func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
 		podUpdates = make(chan podWork, 1)
 		p.podUpdates[uid] = podUpdates
 
+		if kubetypes.IsStaticPod(pod) {
```
Since this chunk is about ensuring that static pods start in FIFO order, please add a comment to this chunk like:

```
// ensure that static pods start in the order they are received by UpdatePod
```

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -734,10 +751,75 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+	defer p.podLock.Unlock()
+	if !p.isStaticPodAllowedToStart(pod) {
+		p.workQueue.Enqueue(pod.UID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
+		if status := p.podSyncStatuses[pod.UID]; status != nil {
```
Instead of checking for nil, always use `ok` (there should never be a nil status)

```
if status, ok := p.podSyncStatuses[pod.UID]; ok {
  status.working = false
}
```

smarterclayton replied:
Actually, getting the syncStatus should be moved up to right after the lock:

```
defer p.podLock.Unlock()
status, ok := p.podSyncStatus[pod.UID]
if !ok {
  return false
}
if !p.isStaticPodAllowedToStart(pod.UID, status.fullName) {
  ...
  status.working = false
}
p.allowNextStaticPod(status.fullName)
return true
```

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -734,10 +751,75 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+	fullname := kubecontainer.GetPodFullName(pod)
+	uid, started := p.startedStaticPodsByFullname[fullname]
+	if started {
+		if uid == pod.UID {
```
This would be clearer if it was simply

```
return uid == pod.UID
```

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -734,10 +751,75 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+// pod with the same full name and pod is the first static pod waiting to
+// start.
+func (p *podWorkers) isStaticPodAllowedToStart(pod *v1.Pod) bool {
+	fullname := kubecontainer.GetPodFullName(pod)
```
This method should take fullname and pod UID - we shouldn't be getting pod full name twice in the same loop

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -246,6 +245,8 @@ type podSyncStatus struct {
 	// working is true if a pod worker is currently in a sync method.
 	working bool
+	// fullname of the pod
+	fullname string
```
`fullName`

smarterclayton replied:
Nm, it looks ilke we use `fullname` consistently everywhere else, so use `fullname` here.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -734,10 +751,75 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+		return
+	}
+
+	p.startedStaticPodsByFullname[fullname] = waitingPods[0]
```
This seems like it is in the wrong place - this method already knows what the started pod UID is, so the method should be taking `(fullName, uid)` just like `isStaticPodAllowedToStart` and you should be setting this to

```
p.startedStaticPodsByFullName[fullName] = uid
```

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -734,10 +751,75 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+	}
+
+	p.startedStaticPodsByFullname[fullname] = waitingPods[0]
+	p.waitingToStartStaticPodsByFullname[fullname] = waitingPods[1:]
```
at this point you should be iterating through the list of waitingToStartStaticPods and removing all the duplicate items:

```
var waitingStaticPods []types.UID
for _, waitingUID := range p.waitingToStartStaticPodsByFullName {
  if waitingUID == uid {
    continue
  }
  if _, ok := p.podSyncStatus[waitingUID]; !ok {
    continue
  }
  waitingStaticPods = append(waitingStaticPods, waitingUID)
}
p.waitingToStartStaticPodsByFullName[fullName] = waitingStaticPods
```

That ensures that when allowNextStaticPod exits, both maps are fully up to date and any duplicates are purged

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -901,16 +980,32 @@ func (p *podWorkers) completeTerminatingRuntimePod(pod *v1.Pod) {
+		if p.startedStaticPodsByFullname[fullname] == pod.UID {
+			delete(p.startedStaticPodsByFullname, fullname)
+		}
+		l := 0
```
This whole section can just be deleted if you perform this logic in https://github.com/kubernetes/kubernetes/pull/104743/commits/e77d81267795d914ef7ebab7e13c6a6baf986971#r720357165.  There should always be a pod worker for any "waiting fullname", and if there isn't that's a bug in pod worker lifecycle that has to get fixed.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -947,6 +1040,25 @@ func (p *podWorkers) completeTerminated(pod *v1.Pod) {
+		if p.startedStaticPodsByFullname[fullname] == pod.UID {
+			delete(p.startedStaticPodsByFullname, fullname)
+		}
+		l := 0
```
can be removed

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -1059,6 +1176,25 @@ func (p *podWorkers) removeTerminatedWorker(uid types.UID) {
+	if p.startedStaticPodsByFullname[fullname] == uid {
+		delete(p.startedStaticPodsByFullname, fullname)
+	}
+	l := 0
```
can be removed as per https://github.com/kubernetes/kubernetes/pull/104743/commits/e77d81267795d914ef7ebab7e13c6a6baf986971#r720358696

smarterclayton commented on pkg/kubelet/pod_workers_test.go:
```diff
@@ -640,3 +652,85 @@ func TestKillPodNowFunc(t *testing.T) {
+			// TODO: Do we want to allow non-static pods with the same full name?
+			// Note that it may disable the force deletion of pods.
+			desc:    "non-static pod",
+			pod:     newPod("test", "test"),
```
I expected a copy of this test that verifies if a static and non-static with different uids and the same full name start, both start (one extra test for "static then non-static" and one for "non-static then static" and both should be allowed)

gjkim42 replied:
I am worried that they will result in a non-static pod and a mirror pod with the same full name in the apiserver so that we may not able to designate one of them - as all kubernetes APIs are based on name / namespace key.

(not sure even it is possible for apiserver(etcd) to allow two pods with the same full name)

smarterclayton commented on pkg/kubelet/pod_workers_test.go:
```diff
@@ -640,3 +652,85 @@ func TestKillPodNowFunc(t *testing.T) {
+				} else {
+					t.Errorf("Pod should not be allowed")
+				}
+			}
```
Add a new test condition here "if allowPodStart was true, the pods in waitingToStartStaticPodsByFullName[fullname] do not contain the UID of the pod, and all pod UIDs in waitingToStartStaticPodsByFullName[fullname] have valid podSyncStatus entries for their UID

ehashman replied:
Still need this added I think

gjkim42 replied:
I think this comment is outdated. 

We decided to trim any pods that are no longer waiting.
https://github.com/kubernetes/kubernetes/pull/104743#discussion_r722286946

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -734,10 +750,80 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+		return true
+	}
+
+	return uid == waitingPods[0]
```
I realized after reviewing this more that we need to integrate the cleanup logic into the ability to check the first start.  Because waiting cleanup happens on startup check vs task cleanup, we can collapse most of allowNextStaticPod into here and potentially remove that method altogether.

Roughly, when we check waitingPods here we want to find the first UID that exists, trim anything we skipped past.  I think at that point the bulk of maintenance that is currently in `allowNextStaticPod` can be moved into here.

Roughly:

```
waitingPods := p.waitingToStartStaticPodsByFullname[fullname]
if len(waitingPods) == 0 {
  p.startedStaticPodsByFullname[fullname] = uid
  return true
}
for i, waitingUID := range waitingPods {
  if _, ok := p.podSyncStatuses[waitingUID]; !ok {
    continue
  }
  p.waitingToStartStaticPodsByFullname[fullname] = waitingPods[i+1:]
  if waitingUID == uid {
    p.startedStaticPodsByFullname[fullname] = uid
    return true
  }
  return false
}
delete(p.waitingToStartStaticPodsByFullname, fullname)
p.startedStaticPodsByFullname[fullname] = uid
return true
```

This should guarantee that we trim any pods that are no longer waiting.  Does this make sense?

gjkim42 replied:
seems fine. Done.

removed verifying the cleanup in `allowNextPodToStart` part as we don't care tail, is this ok?

smarterclayton commented on pkg/kubelet/pod_workers_test.go:
```diff
@@ -640,3 +652,200 @@ func TestKillPodNowFunc(t *testing.T) {
+			allowed: false,
+		},
+		{
+			desc: "static pod if the static pod is the first pod waiting to start and clean up invalid waiting to start static pods entries",
```
With my proposed change up above we should split this into two tests, one that verifies tail is not cleaned up (because we trim the head instead), and one that verifies a non-existent pod that goes first fails (if you add that test before you add my change, I believe the current code will fail, showing that we need to do cleanup before we pick the first pod)

gjkim42 replied:
I think we don't have to verify tail is not cleaned up (we don't have to protect that logic by the test)

smarterclayton commented:
This is lgtm, but would like one other reviewer to do a pass (@ehashman or someone who has been in this recently).

/approve

but giving the chance for one more set of eyes

gjkim42 commented:
/retest

gjkim42 commented:
/retest

gjkim42 commented:
/assign @ehashman 
For final review

gjkim42 commented:
/cc @ehashman @derekwaynecarr @Random-Liu 
Need final review from pod_worker maintainers :)

ehashman commented on pkg/kubelet/pod_workers.go:
```diff
@@ -497,8 +501,19 @@ func (p *podWorkers) ShouldPodContentBeRemoved(uid types.UID) bool {
-	_, ok := p.terminatingStaticPodFullnames[podFullName]
-	return ok
+	uid, started := p.startedStaticPodsByFullname[podFullName]
+	if !started {
+		return false
+	}
```
Is it not possible for a pod that hasn't been started to be terminating? I'm thinking of the edge case for a static pod that might be in the middle of setup (e.g. an initContainer) and gets removed before it reaches Running.

Maybe it could still be in `waitingToStartStaticPodsByFullname`?

ehashman commented on pkg/kubelet/pod_workers.go:
```diff
@@ -766,10 +783,76 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
+		status.working = false
+		return false
+	}
+	if !p.allowStaticPodStart(status.fullname, pod.UID) {
+		p.workQueue.Enqueue(pod.UID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
+		status.working = false
+		return false
```
If a static pod is never allowed to start, what happens? It appears that we keep requeueing it forever. Is that possible?

gjkim42 replied:
Yes, it is possible if the previous static pod is terminating forever, but I think we are ensuring that the termination requested pod terminates after `terminationGracePeriodSeconds`.

ehashman commented on pkg/kubelet/pod_workers.go:
```diff
@@ -933,16 +1013,17 @@ func (p *podWorkers) completeTerminatingRuntimePod(pod *v1.Pod) {
 		status.working = false
+
+		if p.startedStaticPodsByFullname[status.fullname] == pod.UID {
+			delete(p.startedStaticPodsByFullname, status.fullname)
```
I don't see `waitingToStartStaticPodsByFullname` getting cleared out anywhere when it's != 0... should that happen here or somewhere else?

gjkim42 replied:
There can be more than one waiting to start static pods with the same full name.
So we can't clear `waitingToStartStaticPodsByFullname` here.

ehashman commented on pkg/kubelet/pod_workers_test.go:
```diff
@@ -682,3 +694,189 @@ func TestKillPodNowFunc(t *testing.T) {
+			},
+			waitingToStartStaticPodsByFullname: map[string][]types.UID{
+				"foo_": {
+					types.UID("uid-2"),
+					types.UID("uid-2"),
```
Is the same UID repeated here intentional?

gjkim42 replied:
Maybe I just wanted to make sure that invalid and redundant waiting pods are trimmed.

ehashman commented on test/e2e_node/mirror_pod_grace_period_test.go:
```diff
@@ -53,36 +56,61 @@ var _ = SIGDescribe("MirrorPodWithGracePeriod", func() {
-				time.Sleep(time.Duration(200) * time.Millisecond)
-			}
+			ginkgo.By("wait for the mirror pod to be running for grace period")
+			gomega.Consistently(func() error {
```
Shouldn't this be Eventually if we're waiting? Consistently checks that it's true every time and will fail on the first false if it's not.

gjkim42 commented:
@ehashman 
Could you PTAL again? Thanks!

ehashman reviewed (commented):
Chatted with @smarterclayton about this, I think we should merge and get signal on this from CI ASAP.

Sorry for my delays on review!

/lgtm

liggitt commented:
for anyone subscribed to this PR, a revert is proposed in #107734
